### ISSUE TITLE
Wrong output result  to the IDE consoles  from Tensor Stream 2.3.4 and 2.6.0.  and wrong result in IDEs

### ISSUE THREAD

**Issue by Vadim-Maklakov** [tensorflow/tensorflow#51997]
Hi all!
**System configuration**
OS Platform and Distribution : Linux Debian 10 4.19.0-17-amd64
TensorFlow installed: binary from pypi.org 
TensorFlow version: 2.3.4 and 2.6.0 
Python version:3.7.12
Installed from: compile from source
Bazel version – not using
GCC/Compiler version – not using
CUDA/cuDNN version – not using
GPU model and memory: AMD FX-8350,  16GB RAM

**Bug describing**
I recently encountered an unexpected problem with output to the IDE console. I prefer to use the Spyder 5 IDE. I started learning DL from the book “Deep Learning with Python Second Edition” by Francois Cholet MEAP 2020. When I compiling the code for Chapter 4 " Getting started with neural networks: classification and regression", namely, in 4.1 Classification of film reviews: an example of binary classification and "4.2.1 The Reuters dataset" in the Spyder 5.14 IDE,  the values of loss function and accuracy began   different than in the book. 

I spent a lot of time trying changing the CUDA Toolkit of various versions, changing   the  operating system from Debian to Ubuntu and return back – but the results in the Spyder console always was different then the result  in the book. 

For the sake of interest, I typed this code from this book in Google Collaboration - and a miracle happened - the data from the book and from the console of the Colaboratory notebook coincided. 
After that, I installed various versions of Tensorflow without support for CUDA drivers in different Python 3.7.12 virtual environments.

As a result, the output of the Jupyter  notebooks console in both virtual environments for different versions of Tensorflow coincided with the final data in the book. I also used different IDEs for different virtual environments - Eclipse 2021-06 for TF 2.3.4 and Spyder 5.14 for TF2. 6. 0 - in both IDEs, the console output and the final result were different from the results in the book and in Jupiter notebooks.

Please open the case to fix this bug. 
Best regards, Vadim Maklakov.  

[ch_4_IMDB_issue.tar.gz](https://github.com/tensorflow/tensorflow/files/7165248/ch_4_IMDB_issue.tar.gz)

**Comment by mohantym** [comment]
Hi @Vadim-Maklakov! 
This is not a bug or feature request, for any further queries you may open this issue in [tf discussion forum ](https://discuss.tensorflow.org/)as there is a larger community there.Thank you!

**Comment by Vadim-Maklakov** [comment]
Hi, @mohantym !

The problem isn't in the jupyter notebooks  files.
Please load the "ch_4_IMDB_binary_classification.py" script in the other IDE, for example Spyder or Pydev, run it  and  check    results and plots will be equal results  from the jupyter notebooks  files. In my case I see very distinguished  results from IDE and jupyter notebooks files and I  attach its  in the my first post bellow logs from IDE consoles and plots files from Spyder and Pydev IDE .

I see that loss function and accuracy   values  for fires training model and final training model  differ considerably then jupyter notebooks results.
 
As I understand and Spyder and Jupyter notebooks  use the same IPython  but why  its  give different results in the end in the console and plots?

It turns out in the end that all the code in the future I should write code only Jupyter notebook and  me forbidden  use other IDE for Python?


Best regards, Vadim Maklakov,

**Comment by Vadim-Maklakov** [comment]
Hi, @mohantym !
Do you compile  `ch_4_IMDB_binary_classification.py` script in the other IDE than Jupyter notebook? 
What results are output in the console and the values ​​of variables in IPython in your case?

**Comment by mohantym** [comment]
Hi @sanatmpa1 ,Could you please look into this issue?

**Comment by jvishnuvardhan** [comment]
@Vadim-Maklakov I ran your code in colab and local terminal. I notice same results but don't know whether those results are matching with the results mentioned in the book. 

Here is the stacktrace from terminal

```
(base) vishnuvardhanj-macbookpro:ch_4_IMDB_issue vishnuvardhanj$ python3 ch_4_IMDB_binary_classification.py
Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/imdb.npz
17465344/17464789 [==============================] - 1s 0us/step
<__array_function__ internals>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/keras/datasets/imdb.py:159: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/keras/datasets/imdb.py:160: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])
Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/imdb_word_index.json
1646592/1641221 [==============================] - 0s 0us/step
2021-09-20 17:23:25.962502: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-09-20 17:23:25.962739: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-09-20 17:23:27.823250: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
Epoch 1/20
30/30 [==============================] - 4s 94ms/step - loss: 0.6932 - acc: 0.4978 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 2/20
30/30 [==============================] - 1s 19ms/step - loss: 0.6931 - acc: 0.5095 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 3/20
30/30 [==============================] - 1s 19ms/step - loss: 0.6931 - acc: 0.5032 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 4/20
30/30 [==============================] - 1s 18ms/step - loss: 0.6931 - acc: 0.5041 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 5/20
30/30 [==============================] - 1s 18ms/step - loss: 0.6931 - acc: 0.5051 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 6/20
30/30 [==============================] - 1s 17ms/step - loss: 0.6931 - acc: 0.5078 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 7/20
30/30 [==============================] - 0s 16ms/step - loss: 0.6931 - acc: 0.5079 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 8/20
30/30 [==============================] - 0s 16ms/step - loss: 0.6931 - acc: 0.5086 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 9/20
30/30 [==============================] - 0s 15ms/step - loss: 0.6931 - acc: 0.5040 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 10/20
30/30 [==============================] - 0s 15ms/step - loss: 0.6931 - acc: 0.5026 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 11/20
30/30 [==============================] - 1s 17ms/step - loss: 0.6931 - acc: 0.5047 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 12/20
30/30 [==============================] - 0s 16ms/step - loss: 0.6931 - acc: 0.5060 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 13/20
30/30 [==============================] - 0s 15ms/step - loss: 0.6932 - acc: 0.4959 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 14/20
30/30 [==============================] - 0s 16ms/step - loss: 0.6931 - acc: 0.5047 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 15/20
30/30 [==============================] - 0s 17ms/step - loss: 0.6931 - acc: 0.5035 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 16/20
30/30 [==============================] - 0s 16ms/step - loss: 0.6931 - acc: 0.5054 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 17/20
30/30 [==============================] - 0s 17ms/step - loss: 0.6931 - acc: 0.5037 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 18/20
30/30 [==============================] - 0s 15ms/step - loss: 0.6931 - acc: 0.5050 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 19/20
30/30 [==============================] - 1s 18ms/step - loss: 0.6932 - acc: 0.5002 - val_loss: 0.6932 - val_acc: 0.4947
Epoch 20/20
30/30 [==============================] - 1s 17ms/step - loss: 0.6931 - acc: 0.5046 - val_loss: 0.6932 - val_acc: 0.4947
49/49 [==============================] - 2s 21ms/step - loss: 0.2500 - acc: 0.4964
782/782 [==============================] - 2s 2ms/step - loss: 0.2500 - acc: 0.5000
```

Also, check results from [colab](https://colab.research.google.com/gist/jvishnuvardhan/57d46ad98ed02f06d7bbe568f9891e28/untitled1050.ipynb). I also pasted your code in colab and ran that also. [gist here](https://colab.research.google.com/gist/jvishnuvardhan/adad828d560b01976552912e5e39435a/untitled1051.ipynb)

I notice all three are matching. Not sure about `Spyder IDE`. 

I don't think this is related to bugs/performance. Please post any further questions in [TF Forum](https://discuss.tensorflow.org/tag/help_request) where there is a big community to support each other. Thanks!

**Comment by Vadim-Maklakov** [comment]
Hi, @jvishnuvardhan!
I think that isn't topic for forum and this issue belong Tensorflow developer.
I get the same values ​​as you  above when I executing the script `ch_4_IMDB_binary_classification.py` in the  Python console or using IDE Spyder, Eclipse PyDev or Pycharm. 
This result of both cases -  you and me -  when executing as script or compiling in the  IDE Spyder, Eclipse PyDev or Pycharm.  does not match the result console output in the book  which must have  be this:
```
Epoch 1/20
2021-09-21 16:43:34.579784: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.11
 8/30 [=======>......................] - ETA: 0s - loss: 0.6436 - acc: 0.6494
2021-09-21 16:43:35.770346: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.11
30/30 [==============================] - 3s 41ms/step - loss: 0.5253 - acc: 0.7831 - val_loss: 0.3976 - val_acc: 0.8716
Epoch 2/20
30/30 [==============================] - 1s 22ms/step - loss: 0.3144 - acc: 0.9030 - val_loss: 0.3158 - val_acc: 0.8854
Epoch 3/20
30/30 [==============================] - 1s 22ms/step - loss: 0.2335 - acc: 0.9269 - val_loss: 0.2900 - val_acc: 0.8887
Epoch 4/20
30/30 [==============================] - 1s 22ms/step - loss: 0.1835 - acc: 0.9429 - val_loss: 0.2745 - val_acc: 0.8910
Epoch 5/20
30/30 [==============================] - 1s 22ms/step - loss: 0.1523 - acc: 0.9524 - val_loss: 0.2796 - val_acc: 0.8886
Epoch 6/20
30/30 [==============================] - 1s 22ms/step - loss: 0.1258 - acc: 0.9621 - val_loss: 0.3459 - val_acc: 0.8712
Epoch 7/20
30/30 [==============================] - 1s 22ms/step - loss: 0.1059 - acc: 0.9701 - val_loss: 0.3153 - val_acc: 0.8828
Epoch 8/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0913 - acc: 0.9753 - val_loss: 0.3264 - val_acc: 0.8789
Epoch 9/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0763 - acc: 0.9789 - val_loss: 0.3627 - val_acc: 0.8713
Epoch 10/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0644 - acc: 0.9840 - val_loss: 0.3556 - val_acc: 0.8795
Epoch 11/20
30/30 [==============================] - 1s 21ms/step - loss: 0.0518 - acc: 0.9871 - val_loss: 0.3862 - val_acc: 0.8788
Epoch 12/20
30/30 [==============================] - 1s 21ms/step - loss: 0.0435 - acc: 0.9905 - val_loss: 0.4076 - val_acc: 0.8770
Epoch 13/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0351 - acc: 0.9926 - val_loss: 0.4554 - val_acc: 0.8678
Epoch 14/20
30/30 [==============================] - 1s 21ms/step - loss: 0.0293 - acc: 0.9940 - val_loss: 0.4661 - val_acc: 0.8715
Epoch 15/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0228 - acc: 0.9961 - val_loss: 0.4923 - val_acc: 0.8716
Epoch 16/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0189 - acc: 0.9973 - val_loss: 0.5289 - val_acc: 0.8688
Epoch 17/20
30/30 [==============================] - 1s 21ms/step - loss: 0.0150 - acc: 0.9981 - val_loss: 0.5599 - val_acc: 0.8704
Epoch 18/20
30/30 [==============================] - 1s 22ms/step - loss: 0.0137 - acc: 0.9977 - val_loss: 0.5907 - val_acc: 0.8697
Epoch 19/20
30/30 [==============================] - 1s 23ms/step - loss: 0.0077 - acc: 0.9997 - val_loss: 0.6228 - val_acc: 0.8668
Epoch 20/20
30/30 [==============================] - 1s 20ms/step - loss: 0.0078 - acc: 0.9992 - val_loss: 0.6653 - val_acc: 0.8689
```
I get same result output console  as book  when I run code in the  local Jupyter  notebook with TF2.5.1 GPU edition (see attach bellow) or when I using [Google Colab](https://colab.research.google.com/drive/14Ft90fajq3kuRcAMwMbwboeDyg1k9PyE?usp=sharing) with TF2.6.0 CPU edition

The main problem is that Tensorflow  used the same Ipython backend but  it gives different results for  different applications. I my case only local Jupyter  notebook  and Google Colab get right result, but Python console or using IDE Spyder, Eclipse PyDev or Pycharm get wrong result. In our case and   Python console and Google Colab get wrong result.

[IMDB_issue_TF251.ipynb.zip](https://github.com/tensorflow/tensorflow/files/7203984/IMDB_issue_TF251.ipynb.zip)

**Comment by Vadim-Maklakov** [comment]
Hi all! My banal mistake type in this code
def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.0
-       return results
+    return results 
Problem solved, I am closing this ticket with my banal mistake. 
My apologies for the unnecessary trouble.
